pypesto/sampling/pymc3.py

```diff
@@ -112,11 +105,221 @@ def get_samples(self) -> McmcPtResult:
 
         if trace_x.shape[0] != trace_fval.shape[0] \
                 or trace_x.shape[1] != trace_fval.shape[1] \
-                or trace_x.shape[2] != len(self.problem.x_names):
+                or trace_x.shape[2] != len(self.problem.x_free_indices):
             raise ValueError("Trace dimensions are inconsistent")
 
         return McmcPtResult(
             trace_x=np.array(trace_x),
             trace_fval=np.array(trace_fval),
             betas=np.array([1.] * trace_x.shape[0]),
         )
+
+
+def create_pymc3_model(problem: Problem,
+                       testval: Union[np.ndarray, None] = None,
+                       beta: float = 1., *,
+                       cache_gradients: bool = True,
+                       vectorize: bool = False,
+                       lerp: str = 'convex',
+                       verbose: bool = False):
+        with pm.Model() as model:
+            # Wrap objective in a theno op (applying caching if needed)
+            objective = problem.objective
+            if objective.has_grad and cache_gradients:
+                objective = CachedObjective(objective)
+            log_post_fun = TheanoLogProbability(objective, beta)
+
+            # If a test value is given, check its size
+            x_free_names = [problem.x_names[i] for i in problem.x_free_indices]
+            if testval is not None and len(testval) != len(x_free_names):
+                raise ValueError('The size of the test value must be equal ' \
+                                 'to the number of free parameters')
+
+            # If a test value is given, correct values at the optimization
+            # boundaries moving them just inside the interval.
+            # This is due to the fact that pymc3 maps bounded variables
+            # to the whole real line.
```

Comment:
The problem is that `pymc3` by default transforms random variables with bounded support to random variables defined on the real line by using a sigmoid transformation. We can tell it not to do it, but sampling does not work anyway (possibly because it tries points outside the support, but it is not clear to me without delving further into `pymc3` internals). As a first consequence intervals become open since the boundaries are mapped to the infinities. As a second consequence, the backward transform used by `pymc3` is not guaranteed to map all real values to values inside the original interval (due to floating point errors).